# 01. Import libraries

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Create a path for easier importing
path = r'C:\Users\blim9\Desktop\Career Foundry\Data Immersion\Achievement 4\Instacart Basket Analysis 2022-09-09'

In [3]:
# Import the customers dataset
df_cust = pd.read_csv(os.path.join (path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
df_cust

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

# 02. Wrangling data

In [5]:
# Rename the columns for uniformity and consistency
df_cust.rename(columns = {'First Name': 'first_name', 'Surnam': 'last_name', 'Gender': 'gender', 'STATE': 'state',
                         'Age': 'age', 'n_dependants': 'num_of_dependants'}, inplace = True)

In [7]:
df_cust

user_id first_name last_name  gender           state  age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        num_of_dependants fam_status  income  
0                       3    married  165665  
1                       0     single   59285  
2                       2    married   99568  
3                       0     single   42049  
4                       1    married   40374  
...                   ...        ...     ...  
206204                  1    married  148828  
206205                  3    married  168639  
206206                  2    married   53374  
206207                  1    married   99799  
206208                  1    married   57095  

[206209 rows x 10 columns]

# 03. Data quality and consistency checks

In [8]:
# Check for missing values
df_cust.isnull().sum()

user_id                  0
first_name           11259
last_name                0
gender                   0
state                    0
age                      0
date_joined              0
num_of_dependants        0
fam_status               0
income                   0
dtype: int64

In [9]:
df_cust[df_cust['first_name'].isnull() == True]

user_id first_name last_name  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        num_of_dependants fam_status  income  
53                      2    married   41709  
73                      0     single   82518  
82                      3    married  117099  
99                      1    married  155673  
105                     3    married  151819  
...                   ...        ...     ...  
206038                  3    married   87783  
206044                  2    married  108488  
206090                  1    married   45275  
206162                  0     single   56325  
206171                  2    married   59222  

[11259 rows x 10 columns]

#### Although there are 11,259 rows with missing values for the first name, we can still identify them by their user id

In [10]:
# Check for duplicates
df_cust[df_cust.duplicated()]

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, num_of_dependants, fam_status, income]
Index: []

#### No duplicates were found

In [12]:
# Check for mixed types
for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

first_name


In [13]:
# Change the dataype of first_name to type string (object)
df_cust['first_name'] = df_cust['first_name'].astype('str')

In [14]:
# Check for mixed types again
for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

#### No more mixed type data

# 04. Merge the data

In [15]:
# Import the most up to date file worked on
df_ords_prods = pd.read_pickle(os.path.join (path, '02 Data', 'Prepared Data', 'ords_prods_revised_v2.pkl'))

In [16]:
df_ords_prods.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  new_customer  product_id  add_to_cart_order  \
0                    NaN          True         196                  1   
1                   15.0         False         196                  1   
2                   21.0         False         196                  1   
3                   29.0         False         196                  1   
4                   28.0         False         196                  1   

   reordered  ... prices_range_label     busiest_day    busiest_days  \
0          0  ...  Mid range product  Regularly busy  Regularly busy   
1          1  ...  Mid range product  Regularly busy    Slowest days   
2          1  ...  Mid range product  Regularly busy    Slowest days   
3          1  ...  Mid range product      Least busy    Slowest days   
4          1  ...  Mid range product      Least busy    Slowest days   

   busiest_period_of_day max_order  loyalty_flag avg_price spending_flag  \
0         Average orders        10  New customer  6.367797   Low spender   
1         Average orders        10  New customer  6.367797   Low spender   
2            Most orders        10  New customer  6.367797   Low spender   
3         Average orders        10  New customer  6.367797   Low spender   
4            Most orders        10  New customer  6.367797   Low spender   

  median_last_order   order_frequency_flag  
0              20.5  Non-frequent customer  
1              20.5  Non-frequent customer  
2              20.5  Non-frequent customer  
3              20.5  Non-frequent customer  
4              20.5  Non-frequent customer  

[5 rows x 25 columns]

In [18]:
# Check the datatype for user_id in df_cust
df_cust['user_id'].dtype

dtype('int64')

In [19]:
# Check the datatype for user_id in df_ords_prods
df_ords_prods['user_id'].dtype

dtype('int64')

In [21]:
# Merge the two dataframes together on user_id
df_merged = df_ords_prods.merge(df_cust, on = 'user_id')

In [22]:
df_merged.shape

(32404859, 34)

In [23]:
df_merged.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  new_customer  product_id  add_to_cart_order  \
0                    NaN          True         196                  1   
1                   15.0         False         196                  1   
2                   21.0         False         196                  1   
3                   29.0         False         196                  1   
4                   28.0         False         196                  1   

   reordered  ...   order_frequency_flag  first_name  last_name  gender  \
0          0  ...  Non-frequent customer       Linda     Nguyen  Female   
1          1  ...  Non-frequent customer       Linda     Nguyen  Female   
2          1  ...  Non-frequent customer       Linda     Nguyen  Female   
3          1  ...  Non-frequent customer       Linda     Nguyen  Female   
4          1  ...  Non-frequent customer       Linda     Nguyen  Female   

     state age date_joined num_of_dependants fam_status  income  
0  Alabama  31   2/17/2019                 3    married   40423  
1  Alabama  31   2/17/2019                 3    married   40423  
2  Alabama  31   2/17/2019                 3    married   40423  
3  Alabama  31   2/17/2019                 3    married   40423  
4  Alabama  31   2/17/2019                 3    married   40423  

[5 rows x 34 columns]

In [24]:
# Export the new merged dataframe into a pickle file
df_merged.to_pickle(os.path.join (path, '02 Data', 'Prepared Data', 'ords_prods_cust_final.pkl'))